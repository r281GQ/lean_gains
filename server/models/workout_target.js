module.exports = mongoose => {
  const workoutTargetSchema = new mongoose.Schema({
    user: {
      ref: 'User',
      type: mongoose.Schema.Types.ObjectId,
      required: true
    },
    type: {
      type: String,
      enum: {
        values: ['main', 'rest'],
        message: `Type must be 'main' or 'rest'!`
      }
    },
    isCycledTraining: {
      type: String,
      enum: {
        values: ['cycle', 'fix'],
        message: `Type must be 'cycle' or 'fix'!`
      }
    },
    startDayofTraining: {
      type: Number
    },
    onEveryxDay: Number,
    name: String,
    onDays: {
      type: [
        {
          type: Number,
          min: 1,
          max: 7
        }
      ]
    },
    exercises: [String]
  });

  mongoose.model('WorkoutTarget', workoutTargetSchema);
};
