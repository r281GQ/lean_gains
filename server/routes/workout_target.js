const {
  handleGetWorkoutTarget,
  handlePostWorkoutTarget,
  handlePutWorkoutTarget,
  handleDeleteWorkoutTarget
} = require('./../services/workout_target');
const withAuth = require('./../services/auth_middleware');

module.exports = app => {
  app.get('/api/workouttargets', withAuth, (request, response) => {
    handleGetWorkoutTarget(request)
      .then(workoutTargets => response.status(200).send(workoutTargets))
      .catch(error => response.sendStatus(501));
  });

  app.post('/api/workouttargets', withAuth, (request, response) => {
    handlePostWorkoutTarget(request)
      .then(workoutTarget => response.status(201).send(workoutTarget))
      .catch(error => response.sendStatus(501));
  });

  app.put('/api/workouttargets', withAuth, (request, response) => {
    handlePutWorkoutTarget(request)
      .then(workoutTarget => response.status(200).send(workoutTarget))
      .catch(error => response.sendStatus(501));
  });

  app.delete('/api/workouttargets/:_id', withAuth, (request, response) => {
    handleDeleteWorkoutTarget(request)
      .then(workoutTarget => response.status(200).send(workoutTarget))
      .catch(error => response.sendStatus(501));
  });
};
