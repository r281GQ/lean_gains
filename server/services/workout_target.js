const mongoose = require('mongoose');

const WorkoutTarget = mongoose.model('WorkoutTarget');

const { decorateWithUser } = require('./../utils/service');

const handleGetWorkoutTarget = ({ user }) =>
  new Promise((resolve, reject) =>
    WorkoutTarget.find({ user })
      .then(targets => resolve(targets))
      .catch(error => reject(error)),
  );

const handlePutWorkoutTarget = ({
  body: {
    _id,
    isCycledTraining,
    name,
    type,
    startDayofTraining,
    exercises,
    onEveryxDay,
    onDays,
  },
  user,
}) =>
  new Promise((resolve, reject) =>
    WorkoutTarget.findOneAndUpdate(
      { user, _id },
      {
        $set: {
          isCycledTraining,
          name,
          type,
          exercises,
          startDayofTraining,
          onEveryxDay,
          onDays,
        },
      },
      { new: true },
    )
      .then(item => resolve(item))
      .catch(error => reject(error)),
  );

const handlePostWorkoutTarget = ({ body, user }) =>
  new Promise((resolve, reject) =>
    decorateWithUser('WorkoutTarget')(body, user)
      .save()
      .then(target => resolve(target))
      .catch(error => reject(error)),
  );

const handleDeleteWorkoutTarget = ({ user, params: { _id } }) =>
  new Promise((resolve, reject) =>
    WorkoutTarget.findOneAndRemove({ _id, user })
      .then(target => resolve(target))
      .catch(error => reject(error)),
  );

module.exports = {
  handlePostWorkoutTarget,
  handleGetWorkoutTarget,
  handlePutWorkoutTarget,
  handleDeleteWorkoutTarget,
};
