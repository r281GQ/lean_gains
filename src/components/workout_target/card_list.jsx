import React from 'react';
import {
  List,
  ListItem,
  Card,
  CardHeader,
  CardText
} from 'material-ui';
import * as _ from 'lodash';

import DeleteButton from './../delete_button';
import ModifyButton from './../modify_button';

const CardList = ({
  editLink,
  collection,
  onModalStateChange,
  setSelectedItem
}) => {
  return (
    <List>
      {_.map(collection, item => {
        return (
          <ListItem disabled={true} key={item._id}>
            <Card title={item.name}>
              <CardHeader title={item.name} actAsExpander={true} />

              <CardText expandable={true}>
                {_.map(item.exercises, exercise =>
                  <div key={item.exercises.indexOf(exercise)}>
                    {exercise}
                  </div>
                )}

                <ModifyButton link={`${editLink}${item._id}`} />

                <DeleteButton
                  setSelectedItem={() => setSelectedItem( item._id)}
                  onModalStateChange={onModalStateChange}
                />
              </CardText>
            </Card>
          </ListItem>
        );
      })}
    </List>
  );
};

export default CardList;
