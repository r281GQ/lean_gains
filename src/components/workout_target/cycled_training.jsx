import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form/immutable';
import { TextField, DatePicker } from 'redux-form-material-ui';

const CycledTraining = ({ isCycledTraining, formatDate }) =>
  isCycledTraining === 'cycle'
    ? <div>
        <Field
          name="startDayofTraining"
          component={DatePicker}
          floatingLabelText="Starting date of the training"
          fullWidth={true}
          format={formatDate}
        />
        <Field
          name="onEveryxDay"
          type="number"
          component={TextField}
          fullWidth={true}
          min={1}
          hintText="Happens on every X days"
        />
      </div>
    : null;

CycledTraining.propTypes = {
  isCycledTraining: PropTypes.oneOf(['fix', 'cycle']),
  formatDate: PropTypes.func.isRequired
};

export default CycledTraining;
