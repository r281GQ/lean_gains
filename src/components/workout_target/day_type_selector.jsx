import React from 'react';
import { Field } from 'redux-form/immutable';
import { SelectField } from 'redux-form-material-ui';
import { MenuItem } from 'material-ui';

const DayTypeSelector = () =>
  <div>
    <Field
      name="type"
      component={SelectField}
      fullWidth={true}
      floatingLabelText="Training day or a rest day?"
    >
      <MenuItem value="main" primaryText="Main" />
      <MenuItem value="rest" primaryText="Rest" />
    </Field>
  </div>;

export default DayTypeSelector;
