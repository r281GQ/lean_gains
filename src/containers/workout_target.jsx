import React from "react";
import {
  Field,
  FieldArray,
  reduxForm,
  formValueSelector,
  formValues,
  SubmissionError
} from "redux-form/immutable";
import {
  TextField,
  SelectField,
  Checkbox,
  DatePicker
} from "redux-form-material-ui";
import { MenuItem, FlatButton } from "material-ui";
import { connect } from "react-redux";

import { required } from "./../services/validators";
import { createWorkoutTarget } from "./../store/actionCreators/user_details_action_creators";

//TODO should be able to delete and edit  workouttargers
const WorkoutTarget = ({
  handleSubmit,
  isCycledTraining,
  createWorkoutTarget,
  reset
}) =>
  <div>
    <form
      onSubmit={handleSubmit(
        ({
          isCycledTraining,
          onEveryxDay,
          name,
          type,
          startDayofTraining,
          exercises,
          monday,
          tuesday,
          wednesday,
          thursday,
          friday,
          saturday,
          sunday
        }) => {
          let onDays = _.filter(
            [
              monday ? 1 : undefined,
              tuesday ? 2 : undefined,
              wednesday ? 3 : undefined,
              thursday ? 4 : undefined,
              friday ? 5 : undefined,
              saturday ? 6 : undefined,
              sunday ? 7 : undefined
            ],
            item => item !== undefined
          );

          if (!isCycledTraining && _.isEmpty(onDays)) {
            throw new SubmissionError({
              isCycledTraining:
                "on fixed days you must select at least one day",
              _error: "days"
            });
          }

          if (isCycledTraining && (!startDayofTraining || !onEveryxDay))
            throw new SubmissionError({
              isCycledTraining:
                "on cycled training you must set the starting date and on which days the training should occur",
              _error: "exercise name"
            });
          if (_.isEmpty(exercises) || _.every(exercises, exercise => !exercise))
            throw new SubmissionError({
              isCycledTraining: "you must provide at least one valid exercise",
              _error: "exercise name"
            });

          let workoutTarget = isCycledTraining
            ? { name, type, startDayofTraining, exercises, onEveryxDay }
            : {
                name,
                type,
                exercises,
                onDays
              };
          createWorkoutTarget(workoutTarget);
          reset();
        }
      )}
    >
      <div>
        <Field
          name="isCycledTraining"
          component={SelectField}
          fullWidth={true}
          floatingLabelText="Training type: should it happen on fixed days like every monday or rather on every for example 5 days?"
        >
          <MenuItem value={true} primaryText="Cycled training" />
          <MenuItem value={false} primaryText="On fixed days" />
        </Field>
      </div>
      <div>
        <Field
          name="name"
          floatingLabelText="Workout name"
          hintText="name"
          type="text"
          component={TextField}
          fullWidth={true}
          validate={required("workout must have a name")}
        />
      </div>
      <div>
        <Field
          name="type"
          component={SelectField}
          value="main"
          fullWidth={true}
          floatingLabelText="Will it be on a training day or a rest day?"
        >
          <MenuItem value="main" primaryText="Main" />
          <MenuItem value="rest" primaryText="Rest" />
        </Field>
      </div>
      {isCycledTraining
        ? <div>
            <Field
              name="startDayofTraining"
              component={DatePicker}
              floatingLabelText="starting date of the training"
              fullWidth={true}
              format={(value, name) => (value === "" ? 0 : value)}
            />
            <Field
              name="onEveryxDay"
              type="number"
              component={TextField}
              fullWidth={true}
              min={1}
              hintText="happens on every x days"
            />
          </div>
        : null}

      {!isCycledTraining
        ? <div style={{ position: "relative" }}>
            <Field
              name="monday"
              label="Monday"
              component={Checkbox}
              style={{ float: "left", width: "15%" }}
            />
            <Field
              name="tuesday"
              label="Tuesday"
              component={Checkbox}
              style={{ float: "left", width: "15%" }}
            />
            <Field
              name="wednesday"
              label="Wednesday"
              component={Checkbox}
              style={{ float: "left", width: "15%" }}
            />
            <Field
              name="thursday"
              label="Thursday"
              component={Checkbox}
              style={{ float: "left", width: "15%" }}
            />
            <Field
              name="friday"
              label="Friday"
              component={Checkbox}
              style={{ float: "left", width: "15%" }}
            />
            <Field
              name="saturday"
              label="Saturday"
              component={Checkbox}
              style={{ float: "left", width: "15%" }}
            />
            <Field
              name="sunday"
              label="Sunday"
              component={Checkbox}
              style={{ float: "right", width: "10%" }}
            />
            <div style={{ clear: "both" }} />
          </div>
        : null}
      <div style={{ textAlign: "center" }}>
        <FieldArray
          name="exercises"
          component={({ fields: { map, push, remove } }) => {
            return (
              <div>
                {" "}<FlatButton
                  onTouchTap={() => push(null)}
                  label={`Add exercise`}
                />
                {map((exec, index) =>
                  <div key={index}>
                    <Field
                      name={exec}
                      floatingLabelText="provide an exercise"
                      hintText="exercise name"
                      component={TextField}
                      type="text"
                      fullWidth={true}
                    />
                    <FlatButton
                      style={{ textAlign: "center" }}
                      onTouchTap={() => remove(index)}
                      label={`Remove exercise`}
                    />
                  </div>
                )}{" "}
              </div>
            );
          }}
        />
      </div>
      <div style={{ textAlign: "center" }}>
        <FlatButton type="submit" label={`Create workout`} />
      </div>
    </form>
  </div>;

const mapDispatchToProps = dispatch => ({
  createWorkoutTarget: workoutLog => dispatch(createWorkoutTarget(workoutLog))
});

export default connect(null, mapDispatchToProps)(
  reduxForm({
    form: "create-workout-target",
    shouldValidate: () => true,
    initialValues: { type: "main", isCycledTraining: false }
  })(formValues("isCycledTraining")(WorkoutTarget))
);
