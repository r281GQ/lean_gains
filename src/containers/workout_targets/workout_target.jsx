import React, { PureComponent } from 'react';
import moment from 'moment';
import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import { reduxForm, formValues, SubmissionError } from 'redux-form/immutable';
import { connect } from 'react-redux';
import { required } from './../../services/validators';
import { Map } from 'immutable';
import * as _ from 'lodash'

import {
  createWorkoutTarget,
  updateWorkoutTarget
} from './../../store/actionCreators/user_details_action_creators';
import WorkoutTargetForm from './../../components/workout_target/workout_target_form';

// TODO: div class wrogn on createform
class WorkoutTarget extends PureComponent {
  componentDidMount() {
    const { defaultValue, initialize } = this.props;
    if (defaultValue) {
      initialize(
        defaultValue.withMutations(map =>
          map
            .update('startDayofTraining', value => moment(value).toDate())
            .set(
              'monday',
              defaultValue.get('onDays').find(value => value === 1)
            )
            .set(
              'tuesday',
              defaultValue.get('onDays').find(value => value === 2)
            )
            .set(
              'wednesday',
              defaultValue.get('onDays').find(value => value === 3)
            )
            .set(
              'thursday',
              defaultValue.get('onDays').find(value => value === 4)
            )
            .set(
              'friday',
              defaultValue.get('onDays').find(value => value === 5)
            )
            .set(
              'saturday',
              defaultValue.get('onDays').find(value => value === 6)
            )
            .set(
              'sunday',
              defaultValue.get('onDays').find(value => value === 7)
            )
        )
      );
    } else {
      initialize(Map().set('type', 'main'));
    }
  }

  render() {
    const {
      match,
      handleSubmit,
      isCycledTraining,
      createWorkoutTarget,
      updateWorkoutTarget,
      reset
    } = this.props;
    return (
      <WorkoutTargetForm
        isCycledTraining = {isCycledTraining}
        formatDate={value => (value === '' ? 0 : value)}
        validateWorkoutName={required('Workout must have a name!')}
        submitHandler={handleSubmit(formProps => {
          const onDays = _.filter(
            [
              formProps.get('monday') ? 1 : undefined,
              formProps.get('tuesday') ? 2 : undefined,
              formProps.get('wednesday') ? 3 : undefined,
              formProps.get('thursday') ? 4 : undefined,
              formProps.get('friday') ? 5 : undefined,
              formProps.get('saturday') ? 6 : undefined,
              formProps.get('sunday') ? 7 : undefined
            ],
            item => item !== undefined
          );

          if (
            formProps.get('isCycledTraining') === 'fix' &&
            _.isEmpty(onDays)
          ) {
            throw new SubmissionError({
              isCycledTraining:
                'on fixed days you must select at least one day',
              _error: 'days'
            });
          }

          if (
            formProps.get('isCycledTraining') === 'cycle' &&
            (!formProps.get('startDayofTraining') ||
              !formProps.get('onEveryxDay'))
          )
            throw new SubmissionError({
              isCycledTraining:
                'On cycled training you must set the starting date and on which days the training should occur!',
              _error: 'exercise name'
            });

          if (
            typeof formProps.get('exercises') === 'undefined' ||
            formProps.get('exercises').isEmpty() ||
            formProps.get('exercises').every(exercise => !exercise)
          )
            throw new SubmissionError({
              isCycledTraining: 'You must provide at least one valid exercise!',
              _error: 'exercise name'
            });



          match.params.id
            ? updateWorkoutTarget({
                ...formProps.toJS(),
                startDayofTraining: moment(
                  formProps.get('startDayofTraining')
                ).valueOf(),
                onDays,
                _id: match.params.id
              })
            : createWorkoutTarget({
                ...formProps.toJS(),
                startDayofTraining: moment(
                  formProps.get('startDayofTraining')
                ).valueOf(),
                onDays: formProps.get('isCycledTraining') === 'fix' ?  onDays : []
              });

          reset();
        })}
      />
    );
  }
}


WorkoutTarget.propTypes = {
  defaultValue: ImmutablePropTypes.map,
  updateWorkoutTarget: PropTypes.func.isRequired,
  createWorkoutTarget: PropTypes.func.isRequired,
  isCycledTraining: PropTypes.oneOf(['fix' , 'cycle'])
}

export default connect(null, { createWorkoutTarget, updateWorkoutTarget })(
  reduxForm({
    form: 'workout-target'
  })(formValues('isCycledTraining')(WorkoutTarget))
);
