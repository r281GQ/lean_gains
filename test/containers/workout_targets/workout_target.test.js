import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import {fromJS} from 'immutable';
import {SubmissionError} from 'redux-form/immutable'

import {PureWorkoutTarget} from './../../../src/containers/workout_targets/workout_target';

Date.now = jest.fn(()=>1)
describe('WorkoutTarget test', () => {
  // before(() => {
  //
  // });
  it('should render', () => {
    const componentDidMountSpy = sinon.spy(
      PureWorkoutTarget.prototype,
      'componentDidMount'
    );

    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({}))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
    }

    expect(componentDidMountSpy.calledOnce).toBe(false)
    expect(updateWorkoutTarget.calledOnce).toBe(false)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(false)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should throw error on fix training days and when no days are selected', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({

        isCycledTraining: 'fix'

      }))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
      console.log(e);
      expect(e.errors).toEqual({
        isCycledTraining: 'on fixed days you must select at least one day',
        _error: 'days'
      })
    }

    expect(updateWorkoutTarget.calledOnce).toBe(false)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(false)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should throw error on fix training days and when exercise are given', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({

        isCycledTraining: 'fix',
        monday: true

      }))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
      console.log(e);
      expect(e.errors).toEqual({
        isCycledTraining: 'You must provide at least one valid exercise!',
        _error: 'exercise name'
      })
    }

    expect(updateWorkoutTarget.calledOnce).toBe(false)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(false)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should throw error on cycle training days and whenstarting is absent ', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({

        isCycledTraining: 'cycle'

      }))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
      console.log(e);
      expect(e.errors).toEqual({
        isCycledTraining:
          'On cycled training you must set the starting date and on which days the training should occur!',
        _error: 'exercise name'
      })
    }

    expect(updateWorkoutTarget.calledOnce).toBe(false)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(false)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should throw error on cycle training days and whenstarting is absent ', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({

        isCycledTraining: 'cycle',
        startDayofTraining: 'something'

      }))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
      console.log(e);
      expect(e.errors).toEqual({
        isCycledTraining:
          'On cycled training you must set the starting date and on which days the training should occur!',
        _error: 'exercise name'
      })
    }

    expect(updateWorkoutTarget.calledOnce).toBe(false)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(false)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });


  it('should throw error on cycle training days and whenstarting is absent ', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({

        isCycledTraining: 'cycle',
        onEveryxDay: 4

      }))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
      console.log(e);
      expect(e.errors).toEqual({
        isCycledTraining:
          'On cycled training you must set the starting date and on which days the training should occur!',
        _error: 'exercise name'
      })
    }

    expect(updateWorkoutTarget.calledOnce).toBe(false)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(false)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should throw error on cycle training days and whenstarting is absent ', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({

        isCycledTraining: 'cycle',
        onEveryxDay: 4,
        startDayofTraining: "s"

      }))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
      console.log(e);
      expect(e.errors).toEqual({
        isCycledTraining: 'You must provide at least one valid exercise!',
        _error: 'exercise name'
      })
    }

    expect(updateWorkoutTarget.calledOnce).toBe(false)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(false)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('should throw error on cycle training days and whenstarting is absent ', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
        match = {{params: {id: 2}}}
      />
    );

    try {

      component.instance()._handleFormSubmit(fromJS({

        isCycledTraining: 'cycle',
        onEveryxDay: 4,
        startDayofTraining: new Date(),
        exercises: ["d"]

      }))
    } catch (e) {
      expect(e).toBeInstanceOf(SubmissionError)
      console.log(e);
      expect(e.errors).toEqual({
        isCycledTraining: 'You must provide at least one valid exercise!',
        _error: 'exercise name'
      })
    }

    expect(updateWorkoutTarget.calledOnce).toBe(true)
    expect(createWorkoutTarget.calledOnce).toBe(false)
    expect(resetSpy.calledOnce).toBe(true)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });


  it('should throw error on cycle training days and whenstarting is absent ', () => {


    const handleSubmitSpy = sinon.stub().returns(f=>f);

    const updateWorkoutTarget= sinon.spy();
    const createWorkoutTarget = sinon.spy();
    const submitHandler = sinon.spy();
    const resetSpy = sinon.spy();

    const component = shallow(
      <PureWorkoutTarget
        handleSubmit = {handleSubmitSpy}
        updateWorkoutTarget = {updateWorkoutTarget}
        createWorkoutTarget ={createWorkoutTarget}
        reset = {resetSpy}
        match = {{params: {id: 2}}}
      />
    );

    // try {
    //
    //   component.instance()._handleFormSubmit(fromJS({
    //
    //     isCycledTraining: 'cycle',
    //     onEveryxDay: 4,
    //     startDayofTraining: new Date(),
    //     exercises: ["d"]
    //
    //   }))
    // } catch (e) {
    //   expect(e).toBeInstanceOf(SubmissionError)
    //   console.log(e);
    //   expect(e.errors).toEqual({
    //     isCycledTraining: 'You must provide at least one valid exercise!',
    //     _error: 'exercise name'
    //   })
    // }

    // expect(updateWorkoutTarget.calledOnce).toBe(true)
    // expect(createWorkoutTarget.calledOnce).toBe(false)
    // expect(resetSpy.calledOnce).toBe(true)

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });
});
