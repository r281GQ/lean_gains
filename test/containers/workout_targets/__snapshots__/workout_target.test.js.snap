// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutTarget test should render 1`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on cycle training days and whenstarting is absent  1`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on cycle training days and whenstarting is absent  2`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on cycle training days and whenstarting is absent  3`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on cycle training days and whenstarting is absent  4`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on cycle training days and whenstarting is absent  5`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on cycle training days and whenstarting is absent  6`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on fix training days and when exercise are given 1`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;

exports[`WorkoutTarget test should throw error on fix training days and when no days are selected 1`] = `
<WorkoutTargetForm
  formatDate={[Function]}
  submitHandler={[Function]}
  validateWorkoutName={[Function]}
/>
`;
